File: python/trie.py
# Python 3 program to demonstrate creation and use of Trie
# Longest Common Prefix
ALPHABET_SIZE = 26
indexs = 0
class TrieNode: 
	# constructor 
	def __init__(self): 
		self.isLeaf = False
		self.children = [None]*ALPHABET_SIZE 

# Function to facilitate insertion in Trie 
# If not present, insert the node in the Trie 
def insert(key, root): 
	pCrawl = root 
	for level in range(len(key)): 
		index = ord(key[level]) - ord('a') 
		if pCrawl.children[index] == None: 
			pCrawl.children[index] = TrieNode() 
		pCrawl = pCrawl.children[index] 
	pCrawl.isLeaf = True

# Function to construct Trie 
def constructTrie(arr, n, root): 
	for i in range(n): 
		insert(arr[i], root) 

# Counts and returns number of children of the node 
def countChildren(node): 
	count = 0
	for i in range(ALPHABET_SIZE): 
		if node.children[i] != None: 
			count +=1
			# Keeping track of diversion in the trie 
			global indexs 
			indexs = i 
	return count 
	
# Perform walk on trie and return longest common prefix 
def walkTrie(root): 
	pCrawl = root 
	prefix = "" 
	while(countChildren(pCrawl) == 1 and pCrawl.isLeaf == False): 
		pCrawl = pCrawl.children[indexs] 
		prefix += chr(97 + indexs) 
	return prefix or -1

# Function that returns longest common prefix 
def commonPrefix(arr, n, root): 
	constructTrie(arr, n, root) 
	return walkTrie(root) 

# Example (uncomment)
#arr = ["geeksforgeeks", "geeks", "geek", "geezer"] 
#print(commonPrefix(TrieNode(),4, root)) 

File: C++/Trie.cpp
struct Trie {
    struct Node {
        char ch;
        int is_ending;
        int cnt;                    // Telling number of children of this Node
        Node* ptr[26];

        Node(char c = '\0') : ch(c), is_ending(false), cnt(0) {
            for (int i = 0; i < 26; i++) {
                ptr[i] = nullptr;
            }
        }
    };

    Node* root_ = new Node();
    Trie() {}

    void insert(string s) {
        Node* temp = root_;
        for(char &ch : s) {
            int x = ch - 'a';
            if(temp->ptr[x] == nullptr) {
                temp->ptr[x] = new Node(ch);
                temp->cnt++;
            }
            temp = temp->ptr[x];
        }
        temp->is_ending++;
    }

    bool find(string s) {
        Node* temp = root_;
        for(char &ch : s) {
            int x = ch - 'a';
            if(temp->ptr[x] == nullptr) {
                return false;
            }
            temp = temp->ptr[x];
        }
        return temp->is_ending;
    }

    bool deleter(Node* temp, string s, int level, int len) {
        if(level == len) {
            if(temp->cnt > 0) {
                temp->is_ending--;
                return false;
            } else {
                return true;
            }
        } else {
            int x = s[level] - 'a';
            if(deleter(temp->ptr[x], s, level+1, len)) {
                free(temp->ptr[x]);
                temp->ptr[x] = nullptr;
                temp->cnt--;
            }
            return (!temp->is_ending && temp->cnt == 0);
        }
    }

    void erase(string s) {
        if(!find(s)) {
            return;
        }
        Node* temp = root_;
        deleter(temp, s, 0, s.size());
    }

    void clear(Node* temp) {
        if(temp == nullptr) {
            return;
        }
        for(int i = 0; i < 26; i++) {
            clear(temp->ptr[i]);
        }
        delete(temp);
    }

    void clear() {
        clear(root_);
        root_ = new Node();
    }
};
